File: chap14/run-model14-7.R
library(dplyr)
library(cmdstanr)
library(KernSmooth)
library(loo)

model <- cmdstan_model('model/model14-7.stan')
N_sim <- 10000
D <- 3
b <- c(1.3, -3.1, 0.7)
SD <- 2.5
EPS <- 1e-12

set.seed(123)
N <- 30
X <- cbind(1, matrix(runif(N*(D-1), -3, 3), N, (D-1)))
Mu <- X %*% b

d_res <- lapply(1:N_sim, function(simID) {
  set.seed(simID)
  Y <- rnorm(N, Mu, SD)
  data <- list(N=N, D=D, X=X, Y=Y)
  mess <- capture.output(
    fit <- model$sample(data=data, seed=123, iter_sampling=2500,
                        parallel_chains=4, show_messages=FALSE))
  yp_ms <- fit$draws('yp', format='matrix')
  ge_by_data_point <- sapply(1:N, function(n) {
    dens <- bkde(yp_ms[,n,drop=TRUE])
    f_pred <- approxfun(dens$x, ifelse(dens$y <= EPS, EPS, dens$y), 
                        yleft=EPS, yright=EPS)
    f_true <- function(y) dnorm(y, mean=Mu[n], sd=SD)
    f_ge   <- function(y) f_true(y)*(-log(f_pred(y)))
    ge <- integrate(f_ge, Mu[n]-6*SD, Mu[n]+6*SD)$value
  })
  ge <- mean(ge_by_data_point)
  
  log_lik_ms <- fit$draws('log_lik', format='matrix')
  waic  <- waic(log_lik_ms)$waic/(2*N)
  looic <- loo(log_lik_ms)$looic/(2*N)
  data.frame(ge=ge, waic=waic, looic=looic)
}) %>% bind_rows()


## calculate entropy
# f_true0 <- function(y) dnorm(y, mean=0, sd=SD)  
# f_en <- function(y) f_true0(y)*(-log(f_true0(y)))
# en <- integrate(f_en, -6*SD, 6*SD)$value

## calculate waic by myself
# waic <- function(log_lik_ms) {
#   tr_error_by_data_point <- - log(colMeans(exp(log_lik_ms)))
#   func_var_by_data_point <- colMeans(log_lik_ms^2) - colMeans(log_lik_ms)^2
#   
#   waic_by_data_point <- tr_error_by_data_point + func_var_by_data_point
#   waic <- mean(waic_by_data_point)
#   return(waic)
# }

File: chap14/run-model14-7.py
import cmdstanpy
import numpy as np
import pandas
from scipy import stats
from scipy.stats import norm
from scipy import integrate
import arviz as az
import logging
from cmdstanpy.utils import get_logger
get_logger().setLevel(logging.ERROR)
import warnings
warnings.filterwarnings('ignore')
np.random.seed(123)

model = cmdstanpy.CmdStanModel(stan_file='model/model14-7.stan')
N_sim = 10000
D = 3
b = np.array([1.3, -3.1, 0.7])
SD = 2.5
N = 30
X = np.hstack((np.ones((N,1)), np.random.uniform(-3, 3, (N, D-1))))
Mu = np.matmul(X, b)

res = []
for simID in range(N_sim):
    np.random.seed(simID)
    Y = np.random.normal(Mu, SD, N)
    data = {'N':N, 'D':D, 'X':X, 'Y':Y}
    fit = model.sample(data=data, seed=123, iter_sampling=2500,
                       parallel_chains=4, show_progress=False)
    yp_ms = fit.stan_variable(var='yp')
    ge_by_data_point = []
    for n in range(N):
        f_pred = stats.gaussian_kde(yp_ms[:,n])
        def f_true(y):
            return(norm.pdf(y, Mu[n], SD))
        def f_ge(y):
            return(f_true(y)*(-f_pred.logpdf(y)))
        ge, _ = integrate.nquad(f_ge, [[Mu[n]-6*SD, Mu[n]+6*SD]])
        ge_by_data_point.append(ge)
    ge = np.mean(np.array(ge_by_data_point))
  
    log_lik_ms = fit.stan_variable(var='log_lik')
    waic = - np.mean(az.waic(fit, pointwise=True).waic_i.to_numpy())
    looic = - np.mean(az.loo(fit, pointwise=True).loo_i.to_numpy())
    res.append([ge, waic, looic])

d_res = pandas.DataFrame(res, columns=['ge', 'waic', 'looic'])

## calculate entropy
# def f_true0(y):
#     return(norm.pdf(y, 0, SD))
# 
# def log_f_true0(y):
#     return(norm.logpdf(y, 0, SD))
# 
# def f_en(y):
#     return(f_true0(y)*(-log_f_true0(y)))
# 
# en, _ = integrate.nquad(f_en, [[-6*SD, 6*SD]])

## calculate waic by myself
# def waic_func(log_lik_ms):
#     tr_error_by_data_point = -np.log(np.mean(np.exp(log_lik_ms), axis=0))
#     func_var_by_data_point = np.mean(log_lik_ms**2, axis=0) \
#                        - np.mean(log_lik_ms, axis=0)**2
#     waic_by_data_point = tr_error_by_data_point + func_var_by_data_point
#     waic = np.mean(waic_by_data_point)
#     return(waic)
